geo/src/algorithm/outlier_detection.rs

```diff
@@ -0,0 +1,417 @@
+use std::cmp::Ordering;
+use std::fmt::Display;
+use std::hash::{Hash, Hasher};
+use std::iter::Sum;
+
+use crate::{GeoFloat, MultiPoint, Point};
+
+use rstar::primitives::GeomWithData;
+use rstar::RTree;
+
+/// Calculate the [Local Outlier Factor](https://en.wikipedia.org/wiki/Local_outlier_factor) of a set of points
+///
+/// Citation: Breunig, M., Kriegel, H., Ng, R., and Sander, J. (2000). *LOF: identifying density-based local
+/// outliers.* In ACM Int. Conf. on Management of Data, pages 93-104. doi: [10.1145/335191.335388](https://doi.org/10.1145/335191.335388)
+///
+/// LOF is an unsupervised algorithm that uses local data for anomaly detection.
+///
+/// Outlier vs inlier classification is **highly dependent** on the shape of the data. LOF values < 1
+/// can generally be considered inliers, but e.g. a highly concentrated, uniform dataset could result in
+/// points with a LOF of 1.05 being outliers.
+/// LOF scores should thus be evaluated in the context of the dataset as a whole in order to classify outliers.
+///
+/// If you wish to run multiple outlier detection processes with differing neighbour counts in order
+/// to build up an ensemble for more robust detection, you can use the [`OutlierDetection::prepared_detector`] method, which retains
+/// the spatial index and point set between runs for greater efficiency.
+pub trait OutlierDetection<T>
+where
+    T: GeoFloat,
+{
+    /// The LOF algorithm. `k_neighbours` specifies the number of neighbours to use for local outlier
+    /// classification. The paper linked above (see p. 100) suggests a `k_neighbours` value of 10 - 20
+    /// as a lower bound for "real-world"
+    /// data. The accuracy of the algorithm can be increased (see p. 101) by performing multiple passes
+    /// over the data using a range of `k_neighbours` values, then aggregating these, taking e.g.
+    /// the minimum or the mean.
+    ///
+    /// If `k_neighbours` >= points in the set, input points will be returned with a `score` of 1.
+    /// If there are at least `k_neighbours` duplicate points of an input point, LOF scores can be ∞ or `NaN`.
+    /// It is thus advisable to **deduplicate** or otherwise ensure the uniqueness of the input points.
+    ///
+    /// # Examples
+    ///
+    /// `MultiPoint`:
+    ///
+    /// ```
+    /// use approx::assert_relative_eq;
+    /// use geo::OutlierDetection;
+    /// use geo::{point, MultiPoint};
+    ///
+    /// let v = vec![
+    ///     point!(x: 0.0, y: 0.0),
+    ///     point!(x: 0.0, y: 1.0),
+    ///     point!(x: 3.0, y: 0.0),
+    ///     point!(x: 1.0, y: 1.0),
+    /// ];
+    /// let mp = MultiPoint(v);
+    ///
+    /// let lofscores = mp.outliers(3);
+    ///
+    /// assert_relative_eq!(lofscores[2].score, 3.3333333333333335);
+    /// ```
+    fn outliers(&self, k_neighbours: usize) -> Vec<LofScore<T>>;
+
+    /// Create a prepared outlier detector for ensemble detection operations
+    fn prepared_detector(&self) -> PreparedDetector<T>;
```

Comment:
Nice! Maybe mention in the docs that you'd want to do this for efficiency sake. It's implied by use of the word "prepared" if you know what that means, but is that kind of obscure?

Maybe something that mentions "efficiency" like:
>  /// A PreparedDetector can efficiently recompute outliers with different k_neigbhors value.